images/haproxy/push-cross.sh

```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+# Copyright 2019 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# builds and pushes the haproxy image for all architectures
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# cd to the repo root
+REPO_ROOT=$(git rev-parse --show-toplevel)
+cd "${REPO_ROOT}"
+
+# cd to the image
+cd ./images/haproxy
+
+IMAGE="${IMAGE:-kindest/haproxy}"
+TAG="${TAG:-2.0.0-alpine}"
+BASE="haproxy:${TAG}"
+
+# tag arch, manifest architecture, variant
+ARCHES=(
+  "amd64,amd64,"
+  "arm32v6,arm,v6"
```

Comment:
is this comma a typo or intended? `arm,v6` or `armv6`

Reply:
docker manifest is a mess, this is intended. see the code below
this is:
- `arm32v6` when pulling the matching base image
- arm for the `architecture` in the manifest
- v6 for the variant
The comment above the array references this, but it's all kinda manifest internal stuff. I didn't want to inline an explanation of manifests 😅

Reply:
these are basically tuple (tagged_architecture,architecture,variant)
then we make another one below with (image_we_build,tagged_architecture,architecture,variant)

Reply:
do we need add arm v7?

Reply:
Alpine doesn't support v7 last I checked, bit arm v6 images / binaries work on v7. I'm pretty sure our only "users" so far are on arm64v8

Reply:
```
 {                                                                                                              
    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",                                         
    "size": 946,
    "digest": "sha256:35d94f72942e907a56cc201a700e8f2378a204f40b641c55969fabbd259d1ca7",
    "platform": {                                                                                                
      "architecture": "arm",                                                                                     
      "os": "linux",                                                                                             
      "variant": "v7"                                                                                            
    }                                                                                                            
  },       
```
it has arm v7 manifest.

Reply:
I confirmed that all the other kubernetes/kubernetes images only have v6 and v8, we can revisit later if that changes.